### ISSUE TITLE
This page can't load Google Maps correctly.

### ISSUE THREAD

**Issue by davwheat** [TGRHavoc/live_map-interface#17]
Looks like Google is cracking down on Maps API key-less use.

https://cloud.google.com/maps-platform/user-guide/account-changes/

**Comment by TGRHavoc** [comment]
Are you sure this is an actual issue?

I've got limited access to the internet at the minute but, I've checked the two websites that I know about and the maps is still functioning as intended (I'll be it with a pop up message). The websites checked: http://map.tgrhavoc.me and http://map.sosarp.net

Afaik, Google have been trying to crack down on keyless API calls for a while now, hence the hack implemented in the commit 1f2f8737

**Comment by davwheat** [comment]
I've had mine showing the error before I added my API Key.

**Comment by davwheat** [comment]
I'm mid-way through a commit on my fork to add an API key setting into the config, and if it's left empty it'll use the hack anyway.

**Comment by davwheat** [comment]
The two you've sent display this issue too https://puu.sh/AE6Gr/a47f74b822.png

They don't allow you to switch between the 4 map types

**Comment by TGRHavoc** [comment]
@davwheat

Hmmm... I don't know if this is something I want implemented in the long run. I'd rather the users not have to set up and pay Google to use the map. I'd rather it be free. I might have to migrate to an open-source alternative or, see if I can't implement another hack to allow users to switch the map types without the need for Google to handle it (may require reloading the map).

As soon as I get a stable internet connection I'll have a depper look into the alternatives and see if I can't keep the map free for users.

**Comment by davwheat** [comment]
They won't have to pay -- you get $200 free credit monthly and people won't go over that.

It's just an option for those who care about it working fully.

**Comment by davwheat** [comment]
https://developers.google.com/maps/billing/understanding-cost-of-use#dynamic-maps
In addition, this shows that it costs 0.007 USD per each map load. For example at 50 map loads a day for a year, it would be $127.75 total for a year. WIth the $200 monthly limit I think it's fine 🙂

**Comment by TGRHavoc** [comment]
Ok. So I've done a bit more research and it seems "OpenLayers" is the free alternative I've been looking for. I don't have much free time on my hands at the moment, working on another project. But, when I get some time I will start working on rewriting the interface to use the opensource alternative.

Yes, it will take some time but I believe the benefits will be worth it. Especially for larger servers that are active 24/7 and have a load of people loading the map up (I know this is unlikely to happen but, I'd rather be prepared for the worse).

OpenLayers: http://openlayers.org

**Comment by davwheat** [comment]
That's a great idea. If you'd like I could look into it and see how easy
it'd be to convert over.

**Comment by TGRHavoc** [comment]
Sure. 

Any help is appreciated. I will also set up a milestone to keep track of the update and a bunch of new issues as to what I will be (hopefully) implementing.

**Comment by JasonO99** [comment]
If I wanted to add an API key, where would I put it?

**Comment by davwheat** [comment]
Just an update on this after the new pricing came into effect:
Google now says that any amount up to 28 000 map loads per day is free! (source: https://cloud.google.com/maps-platform/pricing/sheet/)

A map load is defined as: (source: https://developers.google.com/maps/faq#usage-limits)
* A web page or application displays a map using the Maps JavaScript API.
(AKA whenever the live map is loaded by a player or refreshed. The site also **explicitly** states that

> **After a web page or application loads a map**, or a static map image, or a Street View panorama, **any user interactions with it, such as panning, zooming, or switching map layers, do not generate additional map loads or affect usage limits.**

(Emphasis mine)

##TL;DR

Unless your live map gets loaded more than 28 000 times per day, using an API key SHOULD be fine.

**Comment by milhouz** [comment]
Quick question, so I have a google API key how do I go about implementing this? I saw the config for the work around do I just have to get rid of that whole block? Assistance would be appreciated, haven't done much work with development.

**Comment by TGRHavoc** [comment]
Going to close this. I implemented Leaflet.js in 9ced37749ac2f610afaa21109062dd1dc1f6cac6 and the latest commit on develop branch seems stable enough (just ignore the inaccuracy issue)
